OpenSimRT/RealTime/src/RealTimeAnalysis.cpp

```diff
@@ -0,0 +1,243 @@
+            parameters.ikConstraintsWeight, parameters.ikAccuracy);
+
+    // id
+    inverseDynamics = new InverseDynamics(model, parameters.wrenchParameters);
+
+    // so
+    muscleOptimization = new MuscleOptimization(
+            model, parameters.muscleOptimizationParameters,
+            parameters.momentArmFunction);
+
+    // jr
+    jointReaction = new JointReaction(model, parameters.wrenchParameters);
+}
+
+bool RealTimeAnalysis::shouldTerminate() { return terminationFlag.load(); }
+
+void RealTimeAnalysis::shouldTerminate(bool flag) { terminationFlag = flag; }
+
+void RealTimeAnalysis::run() {
+    thread acquisitionThread(&RealTimeAnalysis::acquisition, this);
+    thread processingThread(&RealTimeAnalysis::processing, this);
+    acquisitionThread.detach();
+    processingThread.detach();
+}
+
+Vector RealTimeAnalysis::prepareUnfilteredData(
+        const Vector& q,
+        const vector<ExternalWrench::Input>& externalWrenches) const {
+    int m = q.size() + externalWrenches.size() * ExternalWrench::Input::size();
+    if (m == 0) { THROW_EXCEPTION("cannot convert from empty"); }
+
+    Vector v(m);
+    v(0, q.size()) = q;
+    for (int i = 0; i < externalWrenches.size(); ++i) {
+        auto w = externalWrenches[i].toVector();
+        v(q.size() + i * w.size(), w.size()) = w;
+    }
+    return v;
+}
+
+void RealTimeAnalysis::acquisition() {
+    try {
+        while (true) {
+            if (shouldTerminate()) THROW_EXCEPTION("Acquisition terminated.");
+
+            // get data
+            auto acquisitionData = parameters.dataAcquisitionFunction();
+            if (previousAcquisitionTime >= acquisitionData.IkFrame.t) {
+                continue;
+            }
+
+            // update time
+            previousAcquisitionTime = acquisitionData.IkFrame.t;
+
+            // perform ik
+            auto pose = inverseKinematics->solve(acquisitionData.IkFrame);
+
+            // filter
+            auto unfilteredData = prepareUnfilteredData(
+                    pose.q, acquisitionData.ExternalWrenches);
+            auto filteredData = lowPassFilter->filter({pose.t, unfilteredData});
+
+            // push to buffer
+            if (!filteredData.isValid) continue;
+            buffer.add(filteredData);
+        }
+    } catch (exception& e) {
+        cout << e.what() << endl;
+
+        // raise termination flag
+        terminationFlag = true;
+
+        // notify buffer to not get stuck in processing thread
+        buffer.externalNotify();
+    }
+}
+
+void RealTimeAnalysis::processing() {
+    try {
+        FilteredData filteredData;
+        Vector am, fm, residuals, reactionWrenchVector;
+        Vector_<SpatialVec> reactionWrenches;
+        while (true) {
+            if (shouldTerminate()) THROW_EXCEPTION("Processing terminated.");
+
+            // get data from buffer
+            auto data = buffer.get(1)[0];
+            filteredData.fromVector(data.t, data.x, data.xDot, data.xDDot,
```

Comment:
To speed up things in data transfer, maybe we can use pointers to point to the new vectors.